More migration edits (#6714)
apollographql/apollo-server: docs/source/migration.mdx

</MultiCodeBlock>

> The `listen` option to `startStandaloneServer` replaces the arguments to the `server.listen()` function. (If you passed no arguments to `server.listen()`, you don't need to specify a `listen` option.) This object has the same type as the [options argument passed to Node's `net.Server.listen`](https://nodejs.org/api/net.html#serverlistenoptions-callback).

Author: 9b80f4daf

In Apollo Server 3, the top-level [`dataSources` constructor option](/apollo-server/data/data-sources#adding-data-sources-to-apollo-server) essentially adds a post-processing step to your app's context function, creating `DataSource` subclasses and adding them to a `dataSources` field on your [`context`](/apollo-server/data/resolvers/#the-context-argument) object. This means the TypeScript types returned by the `context` function and the `context` types received in the resolvers and plugins are _different_. Additionally, this design obfuscates that `DataSource` objects are created once per request (i.e., like the rest of the context object).

In Apollo Server 4, we have removed the `dataSources` constructor option. You can treat `DataSources` like any other part of your `context` object.

> The one way that Apollo Server 3 treats `DataSource`s specially is that immediately after construction, it would invoke `dataSource.initialize({ cache, context })` on each `DataSource`. So in Apollo Server 4, you should just pass `cache` and `context` to the constructor yourself if you need them. In Apollo Server 4, you can find the `cache` as a new readonly field on the `ApolloServer` itself.

Reviewer: ```suggestion
In Apollo Server 3, immediately after constructing each `DataSource` subclass, Apollo Server invokes the `dataSource.initialize({ cache, context })` function on each new `DataSource`. If you need to replicate this behavior in Apollo Server 4, you can pass the `cache` and `context` arguments to each `DataSource` constructor. In Apollo Server 4, you can find `cache` as a new readonly field on `ApolloServer`.
```

> Let's say this is your code that uses a `RESTDataSource`-based `DataSource` with Apollo Server 3:

Reviewer: ```suggestion
For example,  below, we use the `RESTDataSource` class to create a `DataSource` with Apollo Server 3:
```

<MultiCodeBlock>

> ```ts

Reviewer: ```suggestion
```ts {4-14, 21-35}
```

import { RESTDataSource } from 'apollo-datasource-rest';
import { ApolloServer } from 'apollo-server';

class MoviesAPI extends RESTDataSource {
  baseURL = 'https://movies-api.example.com/';

  willSendRequest(request: RequestOptions) {
    request.headers.set('Authorization', this.context.token);
  }

  async getMovie(id: string): Movie {
    return this.get(`movies/${encodeURIComponent(id)}`);
  }
}

interface ContextValue {
  token: string;
  dataSources: {
    moviesAPI: MoviesAPI;
  }
};

const server = new ApolloServer({
  typeDefs,
  resolvers,
  context: ({ req: ExpressRequest }): Omit<ContextValue, 'dataSources'> => {
    return {
      token: getTokenFromRequest(req),
    };
  },
  dataSources: (): ContextValue['dataSources'] => {
    return {
      moviesAPI: new MoviesAPI(),
    };
  },
});

await server.listen();
```

</MultiCodeBlock>

<!-- TODO(AS4): Update the next line before release -->

> Here's how you would write the same code in Apollo Server 4. (Note that `RESTDataSource` has moved to a new package name (or at least it will [by the time AS4 is released](https://github.com/apollographql/apollo-server/issues/6048)).)

Reviewer: (See comment below for more context on this suggestion)

```suggestion
Here's how you would write the same code in Apollo Server 4:
```

<MultiCodeBlock>

> ```ts

Reviewer: Highlighting `ContextValue`'s `dataSources` key, the line where you take the `cache` from server, and when you actually create the dataSources:
```suggestion
```ts { 25-27, 37, 40-42) 
```

> import { RESTDataSource, RESTDataSourceOptions } from '@apollo/datasource-rest';

Reviewer: I think we should continue using the `apollo-datasource-rest` package in this example until  `@apollo/datasource-rest` is ready to go. Otherwise, this example won't work, leading to confusion for those trying to follow along. 
```suggestion
import { RESTDataSource, RESTDataSourceOptions } from 'apollo-datasource-rest';
```

I'd instead suggest moving the above note: 

> Note that in the future, the `apollo-datasource-rest` package is going to be renamed  `@apollo/datasource-rest` [by the time Apollo Server 4 is released](https://github.com/apollographql/apollo-server/issues/6048)).

to below this snippet, so folks still have the heads up that the package name is _going_ to change soon.

Author: I don't think `apollo-datasource-rest` will actually work with AS4 for various technical reasons (basically, it depends on some AS3 stuff that will confuse the TS compiler, and also it expects the `initialize` function to be called). Maybe we should call this out with a big "hey this doesn't quite work yet" at the beginning of the section. We should work on this pretty soon I think. We can link to https://github.com/apollographql/apollo-server/issues/6048 in the callout.

import { ApolloServer } from '@apollo/server';
import { startStandaloneServer } from '@apollo/server/standalone';

class MoviesAPI extends RESTDataSource {
  baseURL = 'https://movies-api.example.com/';
  private token: string;

  constructor(options: { token: string } & RESTDataSourceOptions) {
    super(options);  // this should send `cache` through
    this.token = options.token;
  }

  willSendRequest(request: RequestOptions) {
    request.headers.set('Authorization', this.token);
  }

  async getMovie(id: string): Movie {
    return this.get(`movies/${encodeURIComponent(id)}`);
  }
}

interface ContextValue {
  token: string;
  dataSources: {
    moviesAPI: MoviesAPI;
  }
};

const server = new ApolloServer<ContextValue>({
  typeDefs,
  resolvers,
});

> await startStandaloneServer(server, {

Reviewer: Add a line to separate the two:
```suggestion

await startStandaloneServer(server, {
```

  context: ({ req }) => {
    const token = getTokenFromRequest(req);
    const { cache } = server;
    return {
      token,
      dataSources: {
        moviesAPI: new MoviesAPI({ cache, token }),
      },
    };
  },
});
```

</MultiCodeBlock>

> If you actually need access to the entire context value within your `DataSource` rather than just particular fields, you can make your context value be a `class`, which is allowed to refer to itself via `this` in its constructor:

Reviewer: ```suggestion
If you want to access your entire context's value within your `DataSource`, you can do so by making your context value a `class`(enabling it to refer to itself via `this` in its constructor):
```

<MultiCodeBlock>

> ```ts

Reviewer: Highlighting the `ContextValue` class
```suggestion
```ts {24-36}
```

> import { RESTDataSource, RESTDataSourceOptions } from '@apollo/datasource-rest';

Reviewer: Updating import for the same reason here as mentioned above:
```suggestion
import { RESTDataSource, RESTDataSourceOptions } from 'apollo-datasource-rest';
```

Reviewer: Can't do this -> so resolving.

> 

Reviewer: Adding note calling out this doesn't work yet:
```suggestion

> ⚠️ This feature is in active development and does not currently work as described. 
>
> The Apollo Server 4 alpha doesn't support importing from the `apollo-datasource-rest` package. We intend to [rename](https://github.com/apollographql/apollo-server/issues/6048) the `apollo-datasource-rest` package to `@apollo/datasource-rest` and update it to be fully compatible with Apollo Server 4. 
```

```

### `formatResponse` hook

> In Apollo Server 3, the `formatResponse` hook enables you to transform the structure of GraphQL response objects before they're sent to a client. This hook is called only after an operation executes (not if there is an error like a parse error that prevents execution). It receives the `response` `GraphQLResponse` and the `requestContext` (which does not yet contain the `response`). If it returns a non-null `GraphQLResponse`, that response is used instead of the one it received. This hook is provided as a top-level constructor argument, not as part of the plugin API.

Author: This is a bit subtle: it is called if processing the operation successfully gets to "execution" but not only if it "successfully executes".  ie it's called with errors that happen during execution, but not with parse errors and validation errors and such. (I guess it depends on your definition of "successfully executes".)

There is also a `willSendResponse` plugin hook. This plugin hook is called at a similar place in the lifecycle to `formatResponse`. It receives a `requestContext`; the `GraphQLResponse` is on this object in the `response` field. It may mutate `requestContext.response`. It is called for all requests that get far enough to invoke `requestDidStart` (ie, any request where the JSON body can be properly parsed, etc). Anything that the `formatResponse` hook can do can also be accomplished with the `willSendResponse` plugin hook.

> Apollo Server 4 removes the `formatResponse` hook. We instead recommend using the `willSendResponse` plugin to do the same thing. (Note that [the structure of `GraphQLResponse` has also changed](#graphqlresponse) so that most of its fields are nested under a `result` field.) You can tell if an operation made it to the execution phase by seeing if the result has a `data` field, so if it was important that your `formatResponse` hook is only called if execution occurred, you can check for the existence of this field. (There are some small edge cases here: for example, an error in the coercion of variable values would result in calling `formatResponse` but will not have `data` in the result. If differentiating these edge cases is important to you, open an issue and we'll help.)

Reviewer: Take the information from the old paragraph above this one and combining them together:

```suggestion
Apollo Server 4 removes the `formatResponse` hook. We instead recommend using the `willSendResponse` plugin hook, which enables you to do everything you previously did with `formatResponse`. Apollo Server calls the  `willSendResponse` plugin hook for all requests that get far enough to invoke `requestDidStart` (i.e., requests with a parsable JSON body, etc. ). 

The `willSendResponse` plugin hook receives an operation's `requestContext`, which has a `response` field containing a `GraphQLResponse` object. Note that the `willSendResponse` plugin hook might mutate the `requestContext.response` field. 

> Note that Apollo Server 4 changes the structure of `GraphQLResponse`; [see below for more details](#graphqlresponse). 
```

Is this extra background on `formatResponse` necessary:
> You can tell if an operation made it to the execution phase by seeing if the result has a `data` field, so if it was important that your `formatResponse` hook is only called if execution occurred, you can check for the existence of this field. (There are some small edge cases here: for example, an error in the coercion of variable values would result in calling `formatResponse` but will not have `data` in the result. If differentiating these edge cases is important to you, open an issue and we'll help.)

It doesn't seem to apply to `willSendResponse`, and  if folks are no longer using `formatResponse`, then it feels like maybe this section could be excluded?

Author: willSendResponse gets called in more contexts than formatResponse, so if you care about maintaining the old behavior you have to implement the filter yourself.

You changed "may" to "might", but I meant "may" in the sense of "is allowed to".

Reviewer: Thank for your the clarification, I've updated the below suggestion to incorporate the intentional "may", and the extra info on `willSendResponse` being called more: 

```suggestion
Apollo Server 4 removes the `formatResponse` hook. We instead recommend using the `willSendResponse` plugin hook, which enables you to do everything you previously did with `formatResponse`. The `willSendResponse` plugin hook receives an operation's `requestContext`, which has a `response` field containing the `GraphQLResponse` object. Note that the `willSendResponse` hook may mutate the `requestContext.response` field. 

> Note that Apollo Server 4 changes the structure of `GraphQLResponse`; [see below for more details](#graphqlresponse). 

Apollo Server calls the `willSendResponse` plugin hook for all requests that get far enough to invoke `requestDidStart` (i.e., requests with a parsable JSON body, etc. ). This means that Apollo Server calls the `willSendResponse` hook in more contexts than the previous `formatResponse` hook. 

To only use the `willSendResponse` hook after an operation's "execution" stage (i.e., like the previous `formatResponse` hook), you can make a filter checking for the existence of a `data` field in the result. If an operation has a `data` field in the result, it has made it to the execution phase. Note, there are some edge cases: for example, an error in the coercion of variable values calls `formatResponse` but doesn't have `data` in the result. If differentiating these edge cases is important to you, please open an issue, and we'll help. 
```

For example, if your Apollo Server 3 code used `formatResponse` like this:

```ts
new ApolloServer({
  typeDefs,
  resolvers,
  formatResponse({ response, requestContext }) {
    return {
      ...response,
      extensions: {
        ...(response.extensions),
        hello: 'world',
      },
    },
  },
}),
```

> your Apollo Server 4 code could use `willSendResponse` like this:

Reviewer: ```suggestion
your Apollo Server 4 code can use `willSendResponse`, like so:
```

https://your.server/?query=%7B__typename%7D
```

> (Note that you should configure your health check to send a `apollo-require-preflight: true` request along with this method so that it is not blocked by the [CSRF prevention](/apollo-server/security/cors/#preventing-cross-site-request-forgery-csrf) security feature.)

Reviewer: I think this would be a nice call out: 
```suggestion
> You should also send an `apollo-require-preflight: true` header alongside your health check, so it isn't blocked by the [CSRF prevention](/apollo-server/security/cors/#preventing-cross-site-request-forgery-csrf) feature.
```

Apollo Server 3 depends on the [`graphql-tag`](https://www.npmjs.com/package/graphql-tag) npm package and re-exports its `gql` template literal tag. The `gql` tag is essentially a caching wrapper around `graphql-js`'s parser, and most IDEs know to treat the contents of `gql` strings as GraphQL.

Apollo Server 4 does not depend on the `graphql-tag` library, nor does it export the `gql` tag. If you want to continue using the `gql` tag, you can directly install `graphql-tag` into your app,  then update your import, replacing this line:

```ts
import { gql } from 'apollo-server';
```

with this line:

```ts
import gql from 'graphql-tag';
```

> (Note that `gql` is a named export of `apollo-server` but is the default export of `graphql-tag`.)

Reviewer: ```suggestion
The `gql` tag is the default export for `graphql-tag`.
```

Author: I'm happy to see the wording improved but you're losing the contrast between the two versions here.

> However, no Apollo Server APIs (in v3 or v4) actually require you to pass in the parsed document produced by a `gql` template string; for example, the `typeDefs` constructor option allows your schema to be specified either as a string or as a parsed document. Rather than switching to the `graphql-tag` package, you may wish to simply specify your schema as a string (whether inline in your source file, or in a separate `*.graphql` file read by `fs.readFileSync`). Note that most editors that recognize `gql` tags also treat any string starting with `#graphql` as GraphQL.

Reviewer: ```suggestion
Note that Apollo Server's `typeDefs` constructor option accepts your schema as *either* a string or a parsed document (produced by the `gql` template string). So, in Apollo Server 4, you can switch to using the `graphql-tag` package or write your schema as a string (inline or in separate `*.graphql` files). 

> Many code editors that recognize `gql` tags also treat any strings starting with `#graphql` as GraphQL, enabling syntax highlighting. 
```

Author: Paragraph removed separately

```

### Doubly-escaped `variables` and `extensions` in requests

Apollo Server 3 and 4 both accept `POST` requests with a JSON body.

Apollo Server 3 supports an edge case where the `variables` and
`extensions` fields inside a `POST` request's JSON-encoded body can be JSON-encoded strings.

Apollo Server 4 requires that within a `POST` request's JSON-encoded body, the `variables` and `extensions` fields must be objects (not doubly-encoded).

For example, below is a valid query:

```json
{

>   "query": "{ __typename }", extensions: { "foo": 1 }

Author: ```suggestion
  "query": "{ __typename }", "extensions": { "foo": 1 }
```

}
```

Whereas this query would be invalid:

```json
{

>   "query": "{ __typename }", extensions: "{ \"foo\": 1 }"

Author: ```suggestion
  "query": "{ __typename }", "extensions": "{ \"foo\": 1 }"
```

Apollo Server 4 more consistently handles errors thrown by multiple plugin hooks.  Each error is wrapped in an "Unexpected error handling request" error and invoked using the new `unexpectedErrorProcessingRequest` plugin hook.

### Custom `gateway` implementations

<!-- TODO(AS4): Fix before release -->

> > Note: The current alpha of Apollo Server does not work as an Apollo Gateway. We will fix this before the final v4.0.0 release. You can use this alpha to serve subgraphs, just not Gateways.

Reviewer: ```suggestion
> ⚠️ Note: The alpha version of Apollo Server does **not** work as an Apollo Gateway. You can still use the alpha to serve subgraphs, just not Gateways. We will fix this before the v4.0.0 release.
```
